### DISCUSSION TITLE
Missing documentation: Does anyone know what "add node precedence" does?

### DISCUSSION THREAD

**Discussion by hklarner** [google/or-tools#3789] [Routing (and legacy CP) questions] [answered]
Does anyone use this function and if so what for?

```python
def AddNodePrecedence(self, first_node: int, second_node: int, offset: int):
```

- https://github.com/google/or-tools/blob/6897c64dbf7d18b97d3563d28dc4820e3eb52022/ortools/constraint_solver/routing.h#LL2896C3-L2899C4
- https://developers.google.com/optimization/reference/python/constraint_solver/pywrapcp#addnodeprecedence

**Comment by Mizux** [comment, marked as answer]
https://github.com/google/or-tools/blob/5adb6509ee78ba9dbd82e521c14cb4d45874d7b7/ortools/routing/samples/cvrptw_with_precedences.cc#L177-L195

I would say, it can be useful if you need to deliver few objects at the same "logical" location but the order matter.
(e.g. first deliver foundation, then deliver walls, then the roof).
eg2: first the aquarium, then the fishes).

**Reply by hklarner** [reply]
If I want a node to be visited before another node I usually add a hard constraint on the time or on a stop count dimension. Something like
```python
offset = 10
solver.Add(dimension_stop_count.CumulVar(x) + offset < dimension_stop_count.CumulVar(y))
```

Is this identical to
```python
dimension_stop_count.AddNodePrecedence(first_node=x, second_node=y, offset=10)
```
?

**Reply by Mizux** [reply]
I would say first will modify (i.e. have side effect) on the underlaying CP solver setup while the second may have side effect on the routing gls filter too

here we can see some usage, so it must "disturb" the search.

i.e. Instead of having the CP rejecting the solution candidate, the routing could simply avoid to generate it directly.
So you may have a performance gain on the exploration of the search space.

ref:
```sh
git grep -in "GetNodePrecedences"
ortools/constraint_solver/routing.cc:3075:         dimension->GetNodePrecedences()) {
ortools/constraint_solver/routing.cc:5207:        !dimension->GetNodePrecedences().empty() || num_resource_groups > 1) {
ortools/constraint_solver/routing.cc:5744:      [](RoutingDimension* dim) { return !dim->GetNodePrecedences().empty(); });
ortools/constraint_solver/routing.h:3056:  const std::vector<NodePrecedence>& GetNodePrecedences() const {
ortools/constraint_solver/routing_filters.cc:1325:      dimension.GetNodePrecedences();
ortools/constraint_solver/routing_filters.cc:2192:    const bool has_precedences = !dimension.GetNodePrecedences().empty();
ortools/constraint_solver/routing_lp_scheduling.cc:413:       dimension_.GetNodePrecedences()) {
ortools/constraint_solver/routing_lp_scheduling.cc:2253:       dimension_->GetNodePrecedences()) {
ortools/constraint_solver/routing_lp_scheduling.cc:2459:                      !dimension->GetNodePrecedences().empty()) {
```

**Reply by hklarner** [reply]
Ok, so both approaches are logically equivalent, but
 - approach 2) may be more efficient, and
 - they may find different assignments

**Reply by Regista6** [reply]
Underlying implementation!
https://github.com/google/or-tools/blob/8efe0e73595524423e344fa1fca813602b96d466/ortools/constraint_solver/routing.cc#L3073-L3090
